File: backend/api/search/service.go
/*
 * Copyright 2024 Sowers, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package search

import (
	grpcRequests "bosca.io/api/protobuf"
	grpcContent "bosca.io/api/protobuf/content"
	grpc "bosca.io/api/protobuf/search"
	"bosca.io/pkg/search"
	"context"
)

type service struct {
	grpc.UnimplementedSearchServiceServer

	contentClient grpcContent.ContentServiceClient

	semanticClient search.SemanticClient
	standardClient search.StandardClient
}

func NewService(contentClient grpcContent.ContentServiceClient, semanticClient search.SemanticClient, standardClient search.StandardClient) grpc.SearchServiceServer {
	return &service{
		contentClient:  contentClient,
		semanticClient: semanticClient,
		standardClient: standardClient,
	}
}

func (s *service) Search(ctx context.Context, request *grpc.SearchRequest) (*grpc.SearchResponse, error) {
	if request.Limit == 0 {
		request.Limit = 25
	}
	results, err := s.semanticClient.Search(ctx, s.semanticClient.GetMetadataIndex(), &search.SemanticQuery{
		Query:  request.Query,
		Offset: request.Offset,
		Limit:  request.Limit,
	})
	if err != nil {
		return nil, err
	}
	metadata, err := s.contentClient.GetMetadatas(ctx, &grpcRequests.IdsRequest{Id: results.MetadataIds})
	if err != nil {
		return nil, err
	}
	return &grpc.SearchResponse{
		Metadata: metadata.Metadata,
	}, nil
}

File: backend.workflows/src/util/service.ts
/*
 * Copyright 2024 Sowers, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { IdRequest } from '../generated/protobuf/bosca/requests_pb'
import { Collection } from '../generated/protobuf/bosca/content/collections_pb'
import { useServiceClient } from './util'
import { ContentService } from '../generated/protobuf/bosca/content/service_connect'
import { Metadata } from '../generated/protobuf/bosca/content/metadata_pb'
import { SignedUrl } from '../generated/protobuf/bosca/content/url_pb'
import { Retry } from './retry'

export async function getCollection(id: IdRequest): Promise<Collection> {
  return Retry.execute(10, () => useServiceClient(ContentService).getCollection(id))
}

export async function getMetadata(id: IdRequest): Promise<Metadata> {
  return Retry.execute(10, () => useServiceClient(ContentService).getMetadata(id))
}

export async function getMetadataUploadUrl(id: IdRequest): Promise<SignedUrl> {
  return Retry.execute(10, () => useServiceClient(ContentService).getMetadataUploadUrl(id))
}
